File: R/dynamical_systems.R
#' @title Simulate a Two-Dimensional Damped Harmonic Oscillator with Cubic Dynamics
#' @description This function simulates a two-dimensional cubic system. It is a wrapper for the Python function \code{cubic2d_ode}, which uses a set of ordinary differential equations (ODEs) to simulate the system. Noise can be added to the system.
#' @param n_obs A numeric value specifying the number of observations or time points for which you want to simulate the system.
#' @param dt A numeric value specifying the time step size for the simulation.
#' @param init_conditions A numeric vector specifying the initial conditions for the simulation. It should contain two elements corresponding to the initial values of the two state variables.
#' @param snr A numeric value specifying the signal-to-noise ratio in dB. This is used to add noise to the simulation. If \code{snr} equals 0, no noise is added.
#' @return A matrix where each row corresponds to a time point and each column corresponds to a state variable. The matrix represents the simulated state of the system over time.
#' @export
#' @examples
#' # Simulate a system for 100 time points with a time step size of 0.01,
#' # initial conditions of c(1, 0) and a signal-to-noise ratio of 49 dB.
#' sim <- cubic2d_system(n_obs=5000, 0.01, c(1, 0), 49)
#' @importFrom reticulate py_eval
#' @import reticulate
cubic2d_system <- function(n_obs, dt, init_conditions, snr) {
  py$cubic2d_ode(n_obs, dt, init_conditions, snr)
}
#' @title Simulate a Duffing Oscillator
#' @description This function simulates a duffing system. It is a wrapper for the Python function \code{duffing_ode}, which uses a set of ordinary differential equations (ODEs) to simulate the system. Noise can be added to the system.
#' @param n_obs A numeric value specifying the number of observations or time points for which you want to simulate the system.
#' @param dt A numeric value specifying the time step size for the simulation.
#' @param init_conditions A numeric vector specifying the initial conditions for the simulation. It should contain two elements corresponding to the initial values of the two state variables.
#' @param snr A numeric value specifying the signal-to-noise ratio in dB. This is used to add noise to the simulation. If \code{snr} equals 0, no noise is added.
#' @return A matrix where each row corresponds to a time point and each column corresponds to a state variable. The matrix represents the simulated state of the system over time.
#' @export
#' @examples
#' # Simulate a system for 100 time points with a time step size of 0.01,
#' # initial conditions of c(1, 0) and a signal-to-noise ratio of 49 dB.
#' sim <- duffing_oscillator(n_obs=5000, 0.01, c(1, 0), 49)
#' @importFrom reticulate py_eval
duffing_oscillator <- function(n_obs, dt, init_conditions, snr) {
  py$duffing_ode(n_obs, dt, init_conditions, snr)
}
#' @title Simulate a Two-Dimensional Damped Harmonic Oscillator with Linear Dynamics
#' @description This function simulates a two-dimensional linear system. It is a wrapper for the Python function \code{linear2d_ode}, which uses a set of ordinary differential equations (ODEs) to simulate the system. Noise can be added to the system.
#' @param n_obs A numeric value specifying the number of observations or time points for which you want to simulate the system.
#' @param dt A numeric value specifying the time step size for the simulation.
#' @param init_conditions A numeric vector specifying the initial conditions for the simulation. It should contain two elements corresponding to the initial values of the two state variables.
#' @param snr A numeric value specifying the signal-to-noise ratio in dB. This is used to add noise to the simulation. If \code{snr} equals 0, no noise is added.
#' @return A matrix where each row corresponds to a time point and each column corresponds to a state variable. The matrix represents the simulated state of the system over time.
#' @export
#' @examples
#' # Simulate a system for 100 time points with a time step size of 0.01,
#' # initial conditions of c(1, 0) and a signal-to-noise ratio of 49 dB.
#' sim <- linear2d_system(n_obs=5000, 0.01, c(1, 0), 49)
#' @importFrom reticulate py_eval
linear2d_system <- function(n_obs, dt, init_conditions, snr) {
  py$linear2d_ode(n_obs, dt, init_conditions, snr)
}
#' @title Simulate a Three-Dimensional Linear System
#' @description This function simulates a three-dimensional linear system. It is a wrapper for the Python function \code{linear3d_ode}, which uses a set of ordinary differential equations (ODEs) to simulate the system. Noise can be added to the system.
#' @param n_obs A numeric value specifying the number of observations or time points for which you want to simulate the system.
#' @param dt A numeric value specifying the time step size for the simulation.
#' @param init_conditions A numeric vector specifying the initial conditions for the simulation. It should contain two elements corresponding to the initial values of the two state variables.
#' @param snr A numeric value specifying the signal-to-noise ratio in dB. This is used to add noise to the simulation. If \code{snr} equals 0, no noise is added.
#' @return A matrix where each row corresponds to a time point and each column corresponds to a state variable. The matrix represents the simulated state of the system over time.
#' @export
#' @examples
#' # Simulate a system for 100 time points with a time step size of 0.01,
#' # initial conditions of c(2, 1, 0) and a signal-to-noise ratio of 49 dB.
#' sim <- linear3d_system(n_obs=5000, 0.01, c(2, 1, 0), 49)
#' @importFrom reticulate py_eval
linear3d_system <- function(n_obs, dt, init_conditions, snr) {
  py$linear3d_ode(n_obs, dt, init_conditions, snr)
}
#' @title Simulate the Lorenz Chaotic System
#' @description This function simulates the Lorenz chaotic system. It is a wrapper for the Python function \code{lorenz_ode}, which uses a set of ordinary differential equations (ODEs) to simulate the system. Noise can be added to the system.
#' @param n_obs A numeric value specifying the number of observations or time points for which you want to simulate the system.
#' @param dt A numeric value specifying the time step size for the simulation.
#' @param init_conditions A numeric vector specifying the initial conditions for the simulation. It should contain two elements corresponding to the initial values of the two state variables.
#' @param snr A numeric value specifying the signal-to-noise ratio in dB. This is used to add noise to the simulation. If \code{snr} equals 0, no noise is added.
#' @return A matrix where each row corresponds to a time point and each column corresponds to a state variable. The matrix represents the simulated state of the system over time.
#' @export
#' @examples
#' # Simulate a system for 100 time points with a time step size of 0.01,
#' # initial conditions of c(-8, 8, 7) and a signal-to-noise ratio of 49 dB.
#' sim <- lorenz_system(n_obs=5000, 0.01, c(-8, 8, 7), 49)
#' @importFrom reticulate py_eval
lorenz_system <- function(n_obs, dt, init_conditions, snr) {
  py$lorenz_ode(n_obs, dt, init_conditions, snr)
}
#' @title Simulate the Rossler System
#' @description This function simulates the Rossler system. It is a wrapper for the Python function \code{rossler_ode}, which uses a set of ordinary differential equations (ODEs) to simulate the system. Noise can be added to the system.
#' @param n_obs A numeric value specifying the number of observations or time points for which you want to simulate the system.
#' @param dt A numeric value specifying the time step size for the simulation.
#' @param init_conditions A numeric vector specifying the initial conditions for the simulation. It should contain two elements corresponding to the initial values of the two state variables.
#' @param snr A numeric value specifying the signal-to-noise ratio in dB. This is used to add noise to the simulation. If \code{snr} equals 0, no noise is added.
#' @return A matrix where each row corresponds to a time point and each column corresponds to a state variable. The matrix represents the simulated state of the system over time.
#' @export
#' @examples
#' # Simulate a system for 100 time points with a time step size of 0.01,
#' # initial conditions of c(2, 0, 1) and a signal-to-noise ratio of 49 dB.
#' sim <- rossler_system(n_obs=5000, 0.01, c(2, 0, 1), 49)
#' @importFrom reticulate py_eval
rossler_system <- function(n_obs, dt, init_conditions, snr) {
  py$rossler_ode(n_obs, dt, init_conditions, snr)
}
#' @title Simulate the Van der Pol Oscillator
#' @description This function simulates the Van der Pol Oscillator. It is a wrapper for the Python function \code{vdp_ode}, which uses a set of ordinary differential equations (ODEs) to simulate the system. Noise can be added to the system.
#' @param n_obs A numeric value specifying the number of observations or time points for which you want to simulate the system.
#' @param dt A numeric value specifying the time step size for the simulation.
#' @param init_conditions A numeric vector specifying the initial conditions for the simulation. It should contain two elements corresponding to the initial values of the two state variables.
#' @param mu A numeric value for specifying the damping of the system.
#' @param snr A numeric value specifying the signal-to-noise ratio in dB. This is used to add noise to the simulation. If \code{snr} equals 0, no noise is added.
#' @return A matrix where each row corresponds to a time point and each column corresponds to a state variable. The matrix represents the simulated state of the system over time.
#' @export
#' @examples
#' # Simulate a system for 100 time points with a time step size of 0.01,
#' # initial conditions of c(2, 0), mu = 1.2, and a signal-to-noise ratio of 49 dB.
#' sim <- vdp_oscillator(n_obs=5000, 0.01, c(2, 0), 1.2, 49)
#' @importFrom reticulate py_eval
vdp_oscillator <- function(n_obs, dt, init_conditions, mu, snr) {
  py$vdp_ode(n_obs, dt, init_conditions, mu, snr)
}

File: inst/dynamical_systems.py
"""dynamical_systems.py"""
import numpy as np
from scipy.integrate import odeint


def cubic2d_ode(n_obs, dt, init_conditions, snr):
    """cubic2d system"""
    cubic_matrix = np.array([-0.1, 2, -2, -0.1])

    def cubic2d(x_t, time, parameters):
        """Two-dimensional damped harmonic oscillator with cubic dynamics"""
        return [
            parameters[0] * x_t[0] ** 3 + parameters[1] * x_t[1] ** 3,
            parameters[2] * x_t[0] ** 3 + parameters[3] * x_t[1] ** 3,
        ]

    t_span = np.arange(0, float(n_obs) * dt, dt)
    x_total = odeint(cubic2d, init_conditions, t_span, args=(cubic_matrix,))
    snr_db = 10 ** -(snr / 20)
    # Add noise (dB)
    if snr_db != 0:
        x_init = x_total.copy()
        for i in range(int(x_total.shape[1])):
            x_total[:, i] = x_total[:, i] + snr_db * np.random.normal(
                scale=np.std(x_init[:, i]), size=x_init[:, i].shape
            )

    return x_total


# Duffing Oscillator
# Matrix parameters

def duffing_ode(n_obs, dt, init_conditions, snr):
    """Duffing oscillator"""
    duffing_params = np.array([0.1, 1, 5])


    def duffing(x_t, time, parameters):
        """Duffing Oscillator"""
        return [
            x_t[1],
            (
                (-parameters[0] * x_t[1])
                - (parameters[1] * x_t[0])
                - (parameters[2] * (x_t[0] ** 3))
            ),
        ]
        
    t_span = np.arange(0, float(n_obs) * dt, dt)
    x_total = odeint(duffing, init_conditions, t_span, args=(duffing_params,))
    snr_db = 10 ** -(snr / 20)
    # Add noise (dB)
    if snr_db != 0:
        x_init = x_total.copy()
        for i in range(int(x_total.shape[1])):
            x_total[:, i] = x_total[:, i] + snr_db * np.random.normal(
                scale=np.std(x_init[:, i]), size=x_init[:, i].shape
            )

    return x_total

# Linear2D
def linear2d_ode(n_obs, dt, init_conditions, snr):
    """Linear2D system"""
    linear2d_params = np.array([-0.1, 2, -2, -0.1])


    def linear2d(x_t, time, parameters):
        """Two-dimensional damped harmonic oscillator with linear dynamics"""
        return [
            parameters[0] * x_t[0] + parameters[1] * x_t[1],
            parameters[2] * x_t[0] + parameters[3] * x_t[1],
        ]
    t_span = np.arange(0, float(n_obs) * dt, dt)
    x_total = odeint(linear2d, init_conditions, t_span, args=(linear2d_params,))
    snr_db = 10 ** -(snr / 20)
    # Add noise (dB)
    if snr_db != 0:
        x_init = x_total.copy()
        for i in range(int(x_total.shape[1])):
            x_total[:, i] = x_total[:, i] + snr_db * np.random.normal(
                scale=np.std(x_init[:, i]), size=x_init[:, i].shape
            )

    return x_total

# Linear3D
def linear3d_ode(n_obs, dt, init_conditions, snr):
    """Linear3D system"""
    linear3d_params = np.array([-0.1, 2, -2, -0.1, -0.3])


    def linear3d(x_t, time, parameters):
        """Three-dimensional linear system"""
        return [
            parameters[0] * x_t[0] + parameters[1] * x_t[1],
            parameters[2] * x_t[0] + parameters[3] * x_t[1],
            parameters[4] * x_t[2],
        ]
        
    t_span = np.arange(0, float(n_obs) * dt, dt)
    x_total = odeint(linear3d, init_conditions, t_span, args=(linear3d_params,))
    snr_db = 10 ** -(snr / 20)
    # Add noise (dB)
    if snr_db != 0:
        x_init = x_total.copy()
        for i in range(int(x_total.shape[1])):
            x_total[:, i] = x_total[:, i] + snr_db * np.random.normal(
                scale=np.std(x_init[:, i]), size=x_init[:, i].shape
            )

    return x_total

# Lorenz system
def lorenz_ode(n_obs, dt, init_conditions, snr):
    """Lorenz system"""
    lorenz_params = np.array([10, 28, -8 / 3])


    def lorenz_eq(x_t, t, parameters):
        """Lorenz system"""
        return [
            parameters[0] * (x_t[1] - x_t[0]),
            x_t[0] * (parameters[1] - x_t[2]) - x_t[1],
            x_t[0] * x_t[1] + parameters[2] * x_t[2],
        ]
        
    t_span = np.arange(0, float(n_obs) * dt, dt)
    x_total = odeint(lorenz_eq, init_conditions, t_span, args=(lorenz_params,))
    snr_db = 10 ** -(snr / 20)
    # Add noise (dB)
    if snr_db != 0:
        x_init = x_total.copy()
        for i in range(int(x_total.shape[1])):
            x_total[:, i] = x_total[:, i] + snr_db * np.random.normal(
                scale=np.std(x_init[:, i]), size=x_init[:, i].shape
            )

    return x_total

# Rossler system
def rossler_ode(n_obs, dt, init_conditions, snr):
    """Rossler system"""
    ross_a = 0.2
    ross_b = 0.2
    ross_c = 5.7
    rossler_params = np.array([ross_a, ross_b, ross_c])

    def rossler_eq(x_t, time, parameters):
        """Rossler Equations"""
        return [
            -x_t[1] - x_t[2],
            x_t[0] + (parameters[0] * x_t[1]),
            parameters[1] + (x_t[2] * (x_t[0] - parameters[2])),
        ]
        
    t_span = np.arange(0, float(n_obs) * dt, dt)
    x_total = odeint(rossler_eq, init_conditions, t_span, args=(rossler_params,))
    snr_db = 10 ** -(snr / 20)
    # Add noise (dB)
    if snr_db != 0:
        x_init = x_total.copy()
        for i in range(int(x_total.shape[1])):
            x_total[:, i] = x_total[:, i] + snr_db * np.random.normal(
                scale=np.std(x_init[:, i]), size=x_init[:, i].shape
            )

    return x_total

# VdP
def vdp_ode(n_obs, dt, init_conditions, mu, snr):
    """Expand system"""
    def vdp_osc(x_t, time, mu):
        """Van der Pol Oscillator"""
        return [x_t[1], mu * (1 - x_t[0] ** 2) * x_t[1] - x_t[0]]
    
    t_span = np.arange(0, float(n_obs) * dt, dt)
    x_total = odeint(vdp_osc, init_conditions, t_span, args=(mu,))
    snr_db = 10 ** -(snr / 20)
    # Add noise (dB)
    if snr_db != 0:
        x_init = x_total.copy()
        for i in range(int(x_total.shape[1])):
            x_total[:, i] = x_total[:, i] + snr_db * np.random.normal(
                scale=np.std(x_init[:, i]), size=x_init[:, i].shape
            )

    return x_total

# Lotka-Volterra
def lotka_volterra_ode(n_obs, dt, init_conditions, snr):
    params = np.array([1, -1, -1, 1])
    def lotka_volterra(x, t, a):
        return[
            (a[0] * x[0]) + (a[1] * (x[0] * x[1])),
            (a[2] * x[1]) + (a[3] * (x[0] * x[1])),
        ]
    t_span = np.arange(0, float(n_obs) * dt, dt)
    x_total = odeint(lotka_volterra, init_conditions, t_span, args=(params,))
    snr_db = 10 ** -(snr / 20)
    # Add noise (dB)
    if snr_db != 0:
        x_init = x_total.copy()
        for i in range(int(x_total.shape[1])):
            x_total[:, i] = x_total[:, i] + snr_db * np.random.normal(
                scale=np.std(x_init[:, i]), size=x_init[:, i].shape
            )

    return x_total